pkg/infrastructure/gcp/clusterapi/iam.go

```diff
@@ -102,48 +106,67 @@ func AddServiceAccountRoles(ctx context.Context, projectID, serviceAccountID str
 		return fmt.Errorf("failed to create resourcemanager service: %w", err)
 	}
 
-	policy, err := getPolicy(ctx, service, projectID)
-	if err != nil {
-		return err
-	}
+	backoff := wait.Backoff{
+		Duration: 2 * time.Second,
+		Jitter:   1.0,
+		Steps:    5,
+	}
+	// Get and set the policy in a backoff loop.
+	// If the policy set fails, the policy must be retrieved again via the get before retrying the set.
+	var lastErr error
+	if waitErr := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		policy, err := getPolicy(ctx, service, projectID)
+		if isQuotaExceededError(err) {
+			logrus.Debugf("Failed to get IAM policy, retrying after backoff")
+			return false, nil
+		} else if err != nil {
+			lastErr = fmt.Errorf("failed to get IAM policy, unexpected error: %w", err)
+			return false, err
```

Comment:
This should be the other way around. If you return an error in this callback, that will be the return of `wait.ExponentialBackoff`. The problem is when you _don't_ return an error because you want the callback to be retried:
```suggestion
		if isQuotaExceededError(err) {
		    lastErr = err
			logrus.Debugf("Failed to get IAM policy, retrying after backoff")
			return false, nil
		} else if err != nil {
			return false, fmt.Errorf("failed to get IAM policy, unexpected error: %w", err)
```

Reply:
We only want to retry on the Quota error not on other types or errors, which is what this does. At least this was how Terraform handled it https://github.com/openshift/installer/blob/master/terraform/providers/google/vendor/github.com/hashicorp/terraform-provider-google/google/tpgiamresource/iam.go#L95-L100

Reply:
I know and that hasn't changed with my suggestion. I think you misunderstood what I meant by "save the last error".

Reply:
Yeah that use of lastErr is incorrect, will fix.

Reply:
Thanks @r4f4, I think I got all of the `lastErr` vs `err` problems resolved now.